# Credit Card Project 

The data set will be use in this small project is from the book **Econometric Analysis** by **William Green**. The [credit cared data](https://www.kaggle.com/dansbecker/aer-credit-card-data) is from kaggle. You can also download it from this [here](https://github.com/DrSaadLa/PythonTuts/blob/main/Data/AER_credit_card_data.csv). 

## Data Description

**Card**:    Dummy variable, 1 if application for credit card accepted, 0 if not

**reports**: Number of major derogatory reports

**age**:     Age n years plus twelfths of a year

**income**:  Yearly income (divided by 10,000)

**share**:   Ratio of monthly credit card expenditure to yearly income

**expenditure**: Average monthly credit card expenditure

**owner**:      1 if owns their home, 0 if rent

**selfempl**:   1 if self employed, 0 if not.

**dependents**: 1 + number of dependents

**months**:     Months living at current address

**majorcards**: Number of major credit cards held

**active**:     Number of active credit accounts

## Problem Description

 - In this project, we will mainly focus on **Exploring** the data set and carry some descriptive statistics. Finally we draw some conclusions upon our results

### Import the necessary packages with their aliases

  - Import pandas
  - Import numpy
  - Import matplotlib
  - Import seaborn

In [1]:
import pandas as pd # as= aliase
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

### Load the data set

In [2]:
url = "https://raw.githubusercontent.com/DrSaadLa/PythonTuts/main/Data/credit.csv"

#### Read the The CSV file, and call the data **credit**

In [3]:
cr=pd.read_csv(url)
cr

card  reports       age  income     share  expenditure owner selfemp  \
0     yes        0  37.66667  4.5200  0.033270   124.983300   yes      no   
1     yes        0  33.25000  2.4200  0.005217     9.854167    no      no   
2     yes        0  33.66667  4.5000  0.004156    15.000000   yes      no   
3     yes        0  30.50000  2.5400  0.065214   137.869200    no      no   
4     yes        0  32.16667  9.7867  0.067051   546.503300   yes      no   
...   ...      ...       ...     ...       ...          ...   ...     ...   
1314  yes        0  33.58333  4.5660  0.002146     7.333333   yes      no   
1315   no        5  23.91667  3.1920  0.000376     0.000000    no      no   
1316  yes        0  40.58333  4.6000  0.026513   101.298300   yes      no   
1317  yes        0  32.83333  3.7000  0.008999    26.996670    no     yes   
1318  yes        0  48.25000  3.7000  0.111619   344.157500   yes      no   

      dependents  months  majorcards  active  
0              3      54           1      12  
1              3      34           1      13  
2              4      58           1       5  
3              0      25           1       7  
4              2      64           1       5  
...          ...     ...         ...     ...  
1314           0      94           1      19  
1315           3      12           1       5  
1316           2       1           1       2  
1317           0      60           1       7  
1318           2       2           1       0  

[1319 rows x 12 columns]

## Data Exploration


### Check the data information

In [4]:
cr.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1319 entries, 0 to 1318
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   card         1319 non-null   object 
 1   reports      1319 non-null   int64  
 2   age          1319 non-null   float64
 3   income       1319 non-null   float64
 4   share        1319 non-null   float64
 5   expenditure  1319 non-null   float64
 6   owner        1319 non-null   object 
 7   selfemp      1319 non-null   object 
 8   dependents   1319 non-null   int64  
 9   months       1319 non-null   int64  
 10  majorcards   1319 non-null   int64  
 11  active       1319 non-null   int64  
dtypes: float64(4), int64(5), object(3)
memory usage: 123.8+ KB


### Answer the question Literally

 - How many variables? 12
 
 - How many observations? 1319
 
 - What is the size of the data in memory? 1071.6875 KB
 
 - How many numeric variables?  9
 
 - How many nominal variables?  3

1. Number of variables: 12

2. Number of observations: 1319

3. Size of the data in memory: 1071.6875 KB

4. The number of numeric variables: 9

5. The number of string variables:  3

### Check the head of the data

In [5]:
cr.memory_usage()

Index            128
card           10552
reports        10552
age            10552
income         10552
share          10552
expenditure    10552
owner          10552
selfemp        10552
dependents     10552
months         10552
majorcards     10552
active         10552
dtype: int64

In [38]:
10552*12 +128

126752

In [7]:
cr.head()

card  reports       age  income     share  expenditure owner selfemp  \
0  yes        0  37.66667  4.5200  0.033270   124.983300   yes      no   
1  yes        0  33.25000  2.4200  0.005217     9.854167    no      no   
2  yes        0  33.66667  4.5000  0.004156    15.000000   yes      no   
3  yes        0  30.50000  2.5400  0.065214   137.869200    no      no   
4  yes        0  32.16667  9.7867  0.067051   546.503300   yes      no   

   dependents  months  majorcards  active  
0           3      54           1      12  
1           3      34           1      13  
2           4      58           1       5  
3           0      25           1       7  
4           2      64           1       5

### Check the column names

In [8]:
cr.columns

Index(['card', 'reports', 'age', 'income', 'share', 'expenditure', 'owner',
       'selfemp', 'dependents', 'months', 'majorcards', 'active'],
      dtype='object')

In [9]:
print(cr.columns.name)

None


## Exploratory Data Analysis

In order to do this section please refer to [this tutorial](https://github.com/DrSaadLa/PythonTuts/blob/main/PandasTuts/03.05.%20Pandas%20Operations.ipynb)

## check the summary statistics of the whole data set

In [10]:
cr.describe()

reports          age       income        share  expenditure  \
count  1319.000000  1319.000000  1319.000000  1319.000000  1319.000000   
mean      0.456406    33.213103     3.365376     0.068732   185.057071   
std       1.345267    10.142783     1.693902     0.094656   272.218917   
min       0.000000     0.166667     0.210000     0.000109     0.000000   
25%       0.000000    25.416670     2.243750     0.002316     4.583333   
50%       0.000000    31.250000     2.900000     0.038827   101.298300   
75%       0.000000    39.416670     4.000000     0.093617   249.035800   
max      14.000000    83.500000    13.500000     0.906320  3099.505000   

        dependents       months   majorcards       active  
count  1319.000000  1319.000000  1319.000000  1319.000000  
mean      0.993935    55.267627     0.817286     6.996967  
std       1.247745    66.271746     0.386579     6.305812  
min       0.000000     0.000000     0.000000     0.000000  
25%       0.000000    12.000000     1.000000     2.000000  
50%       1.000000    30.000000     1.000000     6.000000  
75%       2.000000    72.000000     1.000000    11.000000  
max       6.000000   540.000000     1.000000    46.000000

In [39]:
cr.describe().T

count        mean         std       min        25%         50%  \
reports      1319.0    0.456406    1.345267  0.000000   0.000000    0.000000   
age          1319.0   33.213103   10.142783  0.166667  25.416670   31.250000   
income       1319.0    3.365376    1.693902  0.210000   2.243750    2.900000   
share        1319.0    0.068732    0.094656  0.000109   0.002316    0.038827   
expenditure  1319.0  185.057071  272.218917  0.000000   4.583333  101.298300   
dependents   1319.0    0.993935    1.247745  0.000000   0.000000    1.000000   
months       1319.0   55.267627   66.271746  0.000000  12.000000   30.000000   
majorcards   1319.0    0.817286    0.386579  0.000000   1.000000    1.000000   
active       1319.0    6.996967    6.305812  0.000000   2.000000    6.000000   

                    75%         max  
reports        0.000000    14.00000  
age           39.416670    83.50000  
income         4.000000    13.50000  
share          0.093617     0.90632  
expenditure  249.035800  3099.50500  
dependents     2.000000     6.00000  
months        72.000000   540.00000  
majorcards     1.000000     1.00000  
active        11.000000    46.00000

### How many missing values are there if any?

In [11]:
cr.isnull()

card  reports    age  income  share  expenditure  owner  selfemp  \
0     False    False  False   False  False        False  False    False   
1     False    False  False   False  False        False  False    False   
2     False    False  False   False  False        False  False    False   
3     False    False  False   False  False        False  False    False   
4     False    False  False   False  False        False  False    False   
...     ...      ...    ...     ...    ...          ...    ...      ...   
1314  False    False  False   False  False        False  False    False   
1315  False    False  False   False  False        False  False    False   
1316  False    False  False   False  False        False  False    False   
1317  False    False  False   False  False        False  False    False   
1318  False    False  False   False  False        False  False    False   

      dependents  months  majorcards  active  
0          False   False       False   False  
1          False   False       False   False  
2          False   False       False   False  
3          False   False       False   False  
4          False   False       False   False  
...          ...     ...         ...     ...  
1314       False   False       False   False  
1315       False   False       False   False  
1316       False   False       False   False  
1317       False   False       False   False  
1318       False   False       False   False  

[1319 rows x 12 columns]

In [12]:
cr.isnull().sum()

card           0
reports        0
age            0
income         0
share          0
expenditure    0
owner          0
selfemp        0
dependents     0
months         0
majorcards     0
active         0
dtype: int64

### How many unique values in these variables

1. card
2. owner
3. selfemp

Hint: Give the table of each variable.

In [13]:
cr['card'].duplicated().sum()

1317

In [40]:
cr['card'].nunique()

2

In [14]:
cr['owner'].duplicated().sum()

1317

In [15]:
cr['selfemp'].duplicated().sum()

1317

### What is the Average, min, max  Age of Applicants

In [16]:
avg_age = sum(cr['age']) / len(cr['age'])
avg_age

33.21310325496592

In [41]:
cr['age'].mean()

33.21310325496592

In [42]:
cr['age'].min()

0.1666667

In [43]:
cr['age'].max()

83.5

In [17]:
import statistics
statistics.mean(cr['age'])

33.21310325496588

In [18]:
max_age= max(cr['age'])
max_age

83.5

In [19]:
min_age= min(cr['age'])
min_age

0.1666667

### What is the Average, min, max  Income of Applicants

In [20]:
statistics.mean(cr['income'])

3.3653760424564063

In [21]:
max_income= max(cr['income'])
max_income

13.5

In [22]:
min_income= min(cr['income'])
min_income

0.21

### What is the Highest Income, and how old is this person?

Hint: use **idxmax** 

In [23]:
cr['income'].idxmax()

113

In [45]:
cr.iloc[113,2]

45.58333

In [24]:
cr[cr['income'] == 13.5 ]['age']

113    45.58333
Name: age, dtype: float64

In [47]:
cr['age'].iloc[113]

45.58333

In [46]:
cr[cr['income'] == cr['income'].max()]['age']

113    45.58333
Name: age, dtype: float64

### What is the average income by card?

Hint: use groupby function.

In [25]:
cr.groupby('card').mean() ['income']

card
no     3.068509
yes    3.451273
Name: income, dtype: float64

In [50]:
cr.groupby('card')['income'].mean() 

card
no     3.068509
yes    3.451273
Name: income, dtype: float64

### What is the average expenditure by owner and selfemp?

In [51]:
cr.groupby(['owner', 'selfemp'])['expenditure'].mean() 

owner  selfemp
no     no         163.119629
       yes        153.723849
yes    no         219.638257
       yes        145.416488
Name: expenditure, dtype: float64

## Advanced Section

### 1. Change The columns name to uppercase

In [53]:
cr.rename(columns= str.upper, inplace= True)

In [54]:
cr.head()

CARD  REPORTS       AGE  INCOME     SHARE  EXPENDITURE OWNER SELFEMP  \
0  yes        0  37.66667  4.5200  0.033270   124.983300   yes      no   
1  yes        0  33.25000  2.4200  0.005217     9.854167    no      no   
2  yes        0  33.66667  4.5000  0.004156    15.000000   yes      no   
3  yes        0  30.50000  2.5400  0.065214   137.869200    no      no   
4  yes        0  32.16667  9.7867  0.067051   546.503300   yes      no   

   DEPENDENTS  MONTHS  MAJORCARDS  ACTIVE  
0           3      54           1      12  
1           3      34           1      13  
2           4      58           1       5  
3           0      25           1       7  
4           2      64           1       5

In [ ]:
#cr.rename(columns = str.lower, inplace= True)

In [27]:
cr.columns

Index(['card', 'reports', 'age', 'income', 'share', 'expenditure', 'owner',
       'selfemp', 'dependents', 'months', 'majorcards', 'active'],
      dtype='object')

In [28]:
crColumns = cr.columns.str.upper()

In [29]:
crColumns

Index(['CARD', 'REPORTS', 'AGE', 'INCOME', 'SHARE', 'EXPENDITURE', 'OWNER',
       'SELFEMP', 'DEPENDENTS', 'MONTHS', 'MAJORCARDS', 'ACTIVE'],
      dtype='object')

In [30]:
cr_olumns = cr.columns.map(lambda x: x.upper())
cr_olumns

Index(['CARD', 'REPORTS', 'AGE', 'INCOME', 'SHARE', 'EXPENDITURE', 'OWNER',
       'SELFEMP', 'DEPENDENTS', 'MONTHS', 'MAJORCARDS', 'ACTIVE'],
      dtype='object')

### 2. Apply the following functions as follows:

   2.1. min, max, mean, median, std for the numeric variables.
   
   2.2. apply the count function to categorical variables.

In [31]:
cr[['reports', 'age','income', 'share', 'expenditure', 'dependents' , 'months',
   'majorcards', 'active']].mean()

reports          0.456406
age             33.213103
income           3.365376
share            0.068732
expenditure    185.057071
dependents       0.993935
months          55.267627
majorcards       0.817286
active           6.996967
dtype: float64

In [32]:
cr.iloc[:, np.r_[1:6, 8:12 ]] .mean()

reports          0.456406
age             33.213103
income           3.365376
share            0.068732
expenditure    185.057071
dependents       0.993935
months          55.267627
majorcards       0.817286
active           6.996967
dtype: float64

In [33]:
cr.iloc[:, np.r_[1:6, 8:12 ]] .min()

reports        0.000000
age            0.166667
income         0.210000
share          0.000109
expenditure    0.000000
dependents     0.000000
months         0.000000
majorcards     0.000000
active         0.000000
dtype: float64

In [34]:
cr.iloc[:, np.r_[1:6, 8:12 ]] .max()

reports          14.00000
age              83.50000
income           13.50000
share             0.90632
expenditure    3099.50500
dependents        6.00000
months          540.00000
majorcards        1.00000
active           46.00000
dtype: float64

In [35]:
cr.iloc[:, np.r_[1:6, 8:12 ]] .median()

reports          0.000000
age             31.250000
income           2.900000
share            0.038827
expenditure    101.298300
dependents       1.000000
months          30.000000
majorcards       1.000000
active           6.000000
dtype: float64

In [36]:
cr.iloc[:, np.r_[1:6, 8:12 ]] .std()

reports          1.345267
age             10.142783
income           1.693902
share            0.094656
expenditure    272.218917
dependents       1.247745
months          66.271746
majorcards       0.386579
active           6.305812
dtype: float64

In [37]:
cr.iloc[:, np.r_[5:8]] .count()

expenditure    1319
owner          1319
selfemp        1319
dtype: int64

# Best of luck